src/ProductManifold.jl

```diff
@@ -0,0 +1,458 @@
+    return ShapeSpecification{TRanges, TSizes}()
+end
+
+"""
+    ProductArray(shape::ShapeSpecification, data)
+
+An array-based representation for points and tangent vectors on the
+product manifold. `data` contains underlying representation of points
+arranged according to `TRanges` and `TSizes` from `shape`.
+Internal views for each specific sub-point are created and stored in `parts`.
+"""
+struct ProductArray{TM<:ShapeSpecification,T,N,TData<:AbstractArray{T,N},TV<:Tuple} <: AbstractArray{T,N}
+    data::TData
+    parts::TV
+end
+
+# The two-argument version of this constructor is substantially faster than
+# the generic one.
+function ProductArray(M::Type{ShapeSpecification{TRanges, Tuple{Size1, Size2}}}, data::TData) where {TRanges, Size1, Size2, T, N, TData<:AbstractArray{T,N}}
+    views = (SizedAbstractArray{Size1}(view(data, TRanges[1])),
+             SizedAbstractArray{Size2}(view(data, TRanges[2])))
+    return ProductArray{M, T, N, TData, typeof(views)}(data, views)
+end
+
+function ProductArray(M::Type{ShapeSpecification{TRanges, Tuple{Size1, Size2, Size3}}}, data::TData) where {TRanges, Size1, Size2, Size3, T, N, TData<:AbstractArray{T,N}}
+    views = (SizedAbstractArray{Size1}(view(data, TRanges[1])),
+             SizedAbstractArray{Size2}(view(data, TRanges[2])),
+             SizedAbstractArray{Size3}(view(data, TRanges[3])))
+    return ProductArray{M, T, N, TData, typeof(views)}(data, views)
+end
+
+function ProductArray(M::ShapeSpecification{TRanges, TSizes}, data::TData) where {TM, TRanges, TSizes, T, N, TData<:AbstractArray{T,N}}
+    return ProductArray(typeof(M), data)
+end
+
+@doc doc"""
+    prod_point(M::ShapeSpecification, pts...)
+
+Construct a product point from product manifold `M` based on point `pts`
+represented by [`ProductArray`](@ref).
+
+# Example
+To construct a point on the product manifold $S^2 \times \mathbb{R}^2$
+from points on the sphere and in the euclidean space represented by,
+respectively, `[1.0, 0.0, 0.0]` and `[-3.0, 2.0]` you need to construct shape
+specification first. It describes how linear storage of `ProductArray`
+corresponds to array representations expected by `Sphere(2)` and `Euclidean(2)`.
+
+    M1 = Sphere(2)
+    M2 = Euclidean(2)
+    Mshape = Manifolds.ShapeSpecification(M1, M2)
+
+Next, the desired point on the product manifold can be obtained by calling
+`Manifolds.prod_point(Mshape, [1.0, 0.0, 0.0], [-3.0, 2.0])`.
+"""
+function prod_point(M::ShapeSpecification, pts...)
+    data = mapreduce(vcat, pts) do pt
+        reshape(pt, :)
+    end
+    # Array(data) is used to ensure that the data is mutable
+    # `mapreduce` can return `SArray` for some arguments
+    return ProductArray(M, Array(data))
+end
+
+"""
+    proj_product(x::ProductArray, i::Integer)
+
+Project the product array `x` to its `i`th component. A new array is returned.
+"""
+function proj_product(x::ProductArray, i::Integer)
+    return copy(x.parts[i])
+end
```

Comment:
I find the function name a little strange – but I think you discussed that already? For me its not a project but merely an access (but with copy). Btw. Why do you copy? Without that one could just overload `[I]` access for `ProductArray`?

Reply:
I think projections were discussed in the context of embedded manifolds. I've seen such functions from the Cartesian product to one of its factors also called projections but we can use a different name here. Unfortunately, `getindex`/`setindex!` (or `[I]`) is not an option, they are already defined in a different way, as required by broadcasting. I could probably write much more complex broadcasting rules that don't require `getindex` and `setindex!` defined this way but I don't think it's worth the trouble.

I'll remove that `copy`, I was experimenting with some things and didn't remove it.

Reply:
Ah I did not see that difference and was just thinking about ease of use, since (just as a user view again) having a point `p` on a `Productmanifold` a just natural thing to access the `i`th entry would be `p[I]`; but yes I am also not yet so familiar with broadcast rules.

Reply:
So, how do we name this function then?

Reply:
Well, it's a submanifold of the original `PowerManifold`, so maybe `submanifold` could return a specific manifold from a `ProductManifold` and `submanifold_point` could return the point? The downside to this name is that one would expect to be able to provide any number of indices to build a submanifold, not just a single one.

Reply:
Even for `getIndex` I would expect that any sub manifold can be extracted with for example `p[1:2]`, so it would also be a nice idea to have that hear. I like the name `submanifold_point` for its generality.

Reply:
OK, extracting using an arbitrary set of indices is not a problem, I can do that.

Reply:
I just noticed that `submanifold_point` won't fit for extraction of parts of product tangent vectors, cotangent vectors etc. Can we use a name that works in all of these cases, like `submanifold_component` for example? Or something shorter, like just `component`? Or maybe `factor`?

Reply:
Alternatively, we could overload `getproperty` to allow for something like `p.access[i]`.

Reply:
I prefer if we're going to use `submanifold` to get a subset of manifolds that we use something like `submanifold_` here. I think `submanifold_point` _could_ still work, because no manifold is provided, and a `TVector` and `CoTVector` are points on manifolds in their own rights, but since we haven't been using "point" in this general way so far, I can see how it would be confusing. I think `_component` works. If this and power manifold is the only place where we'll be using these functions, it might make sense to use `factor_` instead of `submanifold_`, since that is the more descriptive term.

Reply:
I don't know, I'm not that good at naming things. I think I'll just let you and @kellertuer pick names for this and `prod_point`.

Reply:
Okay, here're my thoughts. Let's keep `submanifold`, since we may have more uses for that function later. I don't think we should name this `submanifold_point`, since one would expect that to return something similar to the original `ProductArray`. Instead, it should be `submanifold_component(s)`, since the user gets a container of parts. I don't know whether the function name should be plural.

Reply:
OK, I've renamed it to `submanifold_component`. Singular seems more appropriate, right now components can be accessed only one at a time since there is no obvious way to handle multi-component access for `ProductArray` (new `ProductArray`? a tuple with views? an array with views?)

BTW, how do we rename `product_point` now?

Reply:
Yay! Picking names! `Supercalifragilistic_expialigetic` (sorry its late). I was liking the name `access(p,index)`or `access(p,indices)` (which would also get in handy if we figure out how to rename that to getindex and get really `p[i]`-things. Otherwise I am also fine with `manifold_component` without plural, since for a single index its a point on a manifold and even for a container of parts, that could be seen as a point on a submanifold that is still a product manifold; the consistent way for both is hence the singular.